<a href="https://colab.research.google.com/github/daniel-234/CarND-Advanced-Lane-Lines/blob/master/MNIST_sample_digits.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Example taken from the book Deep Learning with Python, 2nd Edition
# See https://www.manning.com/books/deep-learning-with-python-second-edition 

This exercise is taken from the Deep Learning with Python book, as a reference. 

"The problem tries to solve classification of grayscale images of handwritten digits (28 x 28 pixels) into their 10 categories (0 through 9)". [book, section 2.1]


In [2]:
from tensorflow import keras
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Let's look at the training data.
train_images.shape
len(train_labels)

# Let's look at  the test data.
test_images.shape
len(test_labels)

# Build the model.
model = keras.models.Sequential([
    keras.layers.Dense(512, activation="relu"),
    keras.layers.Dense(10, activation="softmax")    
])

# Make the model ready for training.
model.compile(
    optimizer="rmsprop", 
    loss="sparse_categorical_crossentropy", 
    metrics="accuracy"
)

# The model expects that data into an array of shape (60,000, 28 * 28), so we reshape it. 
# The images from the dataset store values in the range (0, 255). We need values in the
# range (0, 1), so we convert them to float and divide them by 255.
train_images = train_images.reshape(60000, 28 * 28)
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape(10000, 28 * 28)
test_images = test_images.astype('float32') / 255

# Now we're ready to train the model: we fit the model to its training data.
model.fit(train_images, train_labels, epochs=5, batch_size=128)

11493376/11490434 [==============================] - 0s 0us/step
Epoch 1/5
469/469 [==============================] - 5s 9ms/step - loss: 0.4315 - accuracy: 0.8749
Epoch 2/5
469/469 [==============================] - 4s 9ms/step - loss: 0.1138 - accuracy: 0.9661
Epoch 3/5
469/469 [==============================] - 4s 9ms/step - loss: 0.0694 - accuracy: 0.9799
Epoch 4/5
469/469 [==============================] - 4s 9ms/step - loss: 0.0483 - accuracy: 0.9853
Epoch 5/5
469/469 [==============================] - 4s 9ms/step - loss: 0.0371 - accuracy: 0.9888


In [8]:
# We can now use the model to make predictions on new data.
# Let's take the first 10 samples from the test dataset. 
test_digits = test_images[0:10]
# predictions will be a 2-D array with 10 arrays of 10 probability scores. By taking the highest
# score index for each array, we would have the model prediction for each one of the ten images. 
predictions = model.predict(test_digits)

In [9]:
# Let's find out what is the model prediction for the second image.
predictions[1]

array([1.7463319e-09, 1.7303583e-03, 9.9826604e-01, 9.0738376e-07,
       7.1160552e-14, 9.0211998e-08, 9.9114409e-08, 5.7319507e-15,
       2.4825056e-06, 7.9784412e-14], dtype=float32)

In [10]:
# The second test digit has the highest probability score (9.9999976e-01) at index 2, so
# it must be a 2. Let's use the 'argmax' function to check if it is true. 
predictions[1].argmax()

2

In [11]:
# Let's also check that the test label agrees. 
test_labels[1]

2

In [12]:
# Let's now evaluate the model accuracy over the entire test set. 
test_loss, test_acc = model.evaluate(test_images, test_labels)

print('test_acc: ', test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.0707 - accuracy: 0.9796
test_acc:  0.9796000123023987


As we can see, the test accuracy turns out to be 97.96%, that is quite a bit lower than the training set accuracy. 
That is due to overfitting. 